gnomad_qc/v5/sample_qc/generate_qc_mt.py

```diff
@@ -0,0 +1,271 @@
+    entry_annotations: List[str] = ["GT"],
+    overwrite: bool = False,
+) -> None:
+    """
+    Filter AoU ACAF and exome MTs to QC MT sites, remove samples to exclude, and union MTs.
+
+    .. note::
+        Both AoU MTs have 145192 partitions.
+
+    :param s_to_exclude: Set of sample IDs to exclude from the AoU MatrixTable.
+    :param ht: Table containing the gnomAD QC sites.
+    :param test: If true, filter to the first 2 partitions for testing.
+        Default is False.
+    :param n_partitions: Number of desired partitions for the unioned MatrixTable.
+        Default is 10000.
+    :param entry_annotations: List of entry annotations to keep in the unioned MatrixTable.
+        Default is ["GT"].
+    :param overwrite: Whether to overwrite output data. Default is False.
+    :return: None; function writes unioned MT to temporary path.
+    """
+    logger.info(
+        "Loading AoU ACAF and exome MatrixTables and removing unnecessary annotations..."
+    )
+    acaf_mt = hl.read_matrix_table(
+        f"gs://{AOU_WGS_BUCKET}/acaf_threshold/splitMT/hail.mt"
+    ).select_entries(*entry_annotations)
+    exome_mt = hl.read_matrix_table(
+        f"gs://{AOU_WGS_BUCKET}/exome/splitMT/hail.mt"
+    ).select_entries(*entry_annotations)
+
+    def _filter_aou_mt(
+        mt: hl.MatrixTable,
+        s_to_exclude: hl.expr.SetExpression,
+        ht: hl.Table,
+        test: bool,
+    ) -> hl.MatrixTable:
+        """
+        Filter AoU MT to gnomAD QC MT sites, remove hard filtered or low quality samples, and filter to `adj`.
+
+        .. note::
+            This function uses AoU's `NO_HQ_GENOTYPES` filter to filter to `adj`.
+
+        :param mt: AoU MatrixTable.
+        :param s_to_exclude: Set of sample IDs to exclude from the AoU MatrixTable.
+        :param ht: Table containing the gnomAD QC sites.
+        :param test: If true, filter to the first 2 partitions for testing.
+        :return: Filtered AoU MatrixTable.
+        """
+        if test:
+            logger.info("Filtering AoU MTs to the first 2 partitions for testing...")
+            mt = mt._filter_partitions(range(2))
+
+        # Filter to gnomAD QC sites and use `NO_HQ_GENOTYPES` to filter to `adj`.
+        mt = mt.filter_rows(hl.is_defined(ht[mt.row_key]))
+        mt = mt.filter_rows(~mt.filters.contains("NO_HQ_GENOTYPES"))
+        mt = mt.select_rows()
+
+        # Remove samples to exclude and their variant sites.
+        mt = mt.filter_cols(~s_to_exclude.contains(mt.s))
+        mt = mt.filter_rows(hl.agg.any(mt.GT.is_non_ref()))
+        return mt
+
+    logger.info("Filtering AoU ACAF and exome MatrixTables and unioning (on rows)...")
+    acaf_mt = _filter_aou_mt(acaf_mt, s_to_exclude, ht, test)
+    exome_mt = _filter_aou_mt(exome_mt, s_to_exclude, ht, test)
+    union_mt = acaf_mt.union_rows(exome_mt)
+
+    logger.info("Decreasing partitions and checkpointing...")
+    union_mt = union_mt.naive_coalesce(n_partitions)
+    union_mt.write(
```

Comment:
Move the `naive_coalesce` and write outside of the function, to keep functions from doing read write of the resources